---

title: Title


keywords: fastai
sidebar: home_sidebar



nb_path: "02_ProcessClinvar.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_ProcessClinvar.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;data/clinvar/variant_summary.txt&quot;</span><span class="p">,</span><span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/dan/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">goodStatus</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;criteria provided, single submitter&quot;</span><span class="p">,</span>
              <span class="s2">&quot;criteria provided, multiple submitters, no conflicts&quot;</span><span class="p">,</span>
              <span class="s2">&quot;reviewed by expert panel&quot;</span><span class="p">,</span>
              <span class="s2">&quot;practice guideline&quot;</span><span class="p">,</span>
              <span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">missense</span> <span class="o">=</span> <span class="n">summary</span><span class="p">[(</span><span class="n">summary</span><span class="o">.</span><span class="n">ReviewStatus</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">goodStatus</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">summary</span><span class="o">.</span><span class="n">Type</span> <span class="o">==</span> <span class="s2">&quot;single nucleotide variant&quot;</span><span class="p">)]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">missense</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>#AlleleID</th>
      <th>Type</th>
      <th>Name</th>
      <th>GeneID</th>
      <th>GeneSymbol</th>
      <th>HGNC_ID</th>
      <th>ClinicalSignificance</th>
      <th>ClinSigSimple</th>
      <th>LastEvaluated</th>
      <th>RS# (dbSNP)</th>
      <th>...</th>
      <th>ReviewStatus</th>
      <th>NumberSubmitters</th>
      <th>Guidelines</th>
      <th>TestedInGTR</th>
      <th>OtherIDs</th>
      <th>SubmitterCategories</th>
      <th>VariationID</th>
      <th>PositionVCF</th>
      <th>ReferenceAlleleVCF</th>
      <th>AlternateAlleleVCF</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6</th>
      <td>15044</td>
      <td>single nucleotide variant</td>
      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>
      <td>55572</td>
      <td>FOXRED1</td>
      <td>HGNC:26927</td>
      <td>Pathogenic</td>
      <td>1</td>
      <td>Dec 30, 2019</td>
      <td>267606829</td>
      <td>...</td>
      <td>criteria provided, multiple submitters, no con...</td>
      <td>3</td>
      <td>-</td>
      <td>N</td>
      <td>ClinGen:CA113792,OMIM:613622.0001</td>
      <td>3</td>
      <td>5</td>
      <td>126145284</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>7</th>
      <td>15044</td>
      <td>single nucleotide variant</td>
      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>
      <td>55572</td>
      <td>FOXRED1</td>
      <td>HGNC:26927</td>
      <td>Pathogenic</td>
      <td>1</td>
      <td>Dec 30, 2019</td>
      <td>267606829</td>
      <td>...</td>
      <td>criteria provided, multiple submitters, no con...</td>
      <td>3</td>
      <td>-</td>
      <td>N</td>
      <td>ClinGen:CA113792,OMIM:613622.0001</td>
      <td>3</td>
      <td>5</td>
      <td>126275389</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>10</th>
      <td>15046</td>
      <td>single nucleotide variant</td>
      <td>NM_025152.3(NUBPL):c.166G&gt;A (p.Gly56Arg)</td>
      <td>80224</td>
      <td>NUBPL</td>
      <td>HGNC:20278</td>
      <td>Uncertain significance</td>
      <td>0</td>
      <td>Aug 24, 2020</td>
      <td>200401432</td>
      <td>...</td>
      <td>criteria provided, multiple submitters, no con...</td>
      <td>3</td>
      <td>-</td>
      <td>N</td>
      <td>ClinGen:CA321015,UniProtKB:Q8TB37#VAR_064570,O...</td>
      <td>2</td>
      <td>214885</td>
      <td>32031331</td>
      <td>G</td>
      <td>A</td>
    </tr>
    <tr>
      <th>11</th>
      <td>15046</td>
      <td>single nucleotide variant</td>
      <td>NM_025152.3(NUBPL):c.166G&gt;A (p.Gly56Arg)</td>
      <td>80224</td>
      <td>NUBPL</td>
      <td>HGNC:20278</td>
      <td>Uncertain significance</td>
      <td>0</td>
      <td>Aug 24, 2020</td>
      <td>200401432</td>
      <td>...</td>
      <td>criteria provided, multiple submitters, no con...</td>
      <td>3</td>
      <td>-</td>
      <td>N</td>
      <td>ClinGen:CA321015,UniProtKB:Q8TB37#VAR_064570,O...</td>
      <td>2</td>
      <td>214885</td>
      <td>31562125</td>
      <td>G</td>
      <td>A</td>
    </tr>
    <tr>
      <th>18</th>
      <td>15051</td>
      <td>single nucleotide variant</td>
      <td>NM_139007.3(HFE):c.77-236T&gt;C</td>
      <td>3077</td>
      <td>HFE</td>
      <td>HGNC:4886</td>
      <td>Uncertain significance</td>
      <td>1</td>
      <td>Oct 20, 2020</td>
      <td>28934596</td>
      <td>...</td>
      <td>criteria provided, single submitter</td>
      <td>2</td>
      <td>-</td>
      <td>N</td>
      <td>ClinGen:CA280941,UniProtKB:Q30201#VAR_008730,O...</td>
      <td>3</td>
      <td>12</td>
      <td>26091306</td>
      <td>T</td>
      <td>C</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2018880</th>
      <td>1166722</td>
      <td>single nucleotide variant</td>
      <td>NM_004309.6(ARHGDIA):c.151T&gt;C (p.Tyr51His)</td>
      <td>396</td>
      <td>ARHGDIA</td>
      <td>HGNC:678</td>
      <td>Uncertain significance</td>
      <td>0</td>
      <td>Jan 04, 2021</td>
      <td>-1</td>
      <td>...</td>
      <td>criteria provided, single submitter</td>
      <td>1</td>
      <td>-</td>
      <td>N</td>
      <td>-</td>
      <td>2</td>
      <td>1177422</td>
      <td>81869780</td>
      <td>A</td>
      <td>G</td>
    </tr>
    <tr>
      <th>2018881</th>
      <td>1166723</td>
      <td>single nucleotide variant</td>
      <td>NM_000132.4(F8):c.6123G&gt;C (p.Gln2041His)</td>
      <td>2157</td>
      <td>F8</td>
      <td>HGNC:3546</td>
      <td>Uncertain significance</td>
      <td>0</td>
      <td>Jan 04, 2021</td>
      <td>-1</td>
      <td>...</td>
      <td>criteria provided, single submitter</td>
      <td>1</td>
      <td>-</td>
      <td>N</td>
      <td>-</td>
      <td>2</td>
      <td>1177425</td>
      <td>154129710</td>
      <td>C</td>
      <td>G</td>
    </tr>
    <tr>
      <th>2018882</th>
      <td>1166723</td>
      <td>single nucleotide variant</td>
      <td>NM_000132.4(F8):c.6123G&gt;C (p.Gln2041His)</td>
      <td>2157</td>
      <td>F8</td>
      <td>HGNC:3546</td>
      <td>Uncertain significance</td>
      <td>0</td>
      <td>Jan 04, 2021</td>
      <td>-1</td>
      <td>...</td>
      <td>criteria provided, single submitter</td>
      <td>1</td>
      <td>-</td>
      <td>N</td>
      <td>-</td>
      <td>2</td>
      <td>1177425</td>
      <td>154901435</td>
      <td>C</td>
      <td>G</td>
    </tr>
    <tr>
      <th>2018901</th>
      <td>1166863</td>
      <td>single nucleotide variant</td>
      <td>NM_173630.4(RTTN):c.4438C&gt;T (p.Leu1480Phe)</td>
      <td>25914</td>
      <td>RTTN</td>
      <td>HGNC:18654</td>
      <td>Uncertain significance</td>
      <td>0</td>
      <td>Mar 18, 2020</td>
      <td>-1</td>
      <td>...</td>
      <td>criteria provided, single submitter</td>
      <td>1</td>
      <td>-</td>
      <td>N</td>
      <td>-</td>
      <td>2</td>
      <td>1177441</td>
      <td>67742714</td>
      <td>G</td>
      <td>A</td>
    </tr>
    <tr>
      <th>2018902</th>
      <td>1166863</td>
      <td>single nucleotide variant</td>
      <td>NM_173630.4(RTTN):c.4438C&gt;T (p.Leu1480Phe)</td>
      <td>25914</td>
      <td>RTTN</td>
      <td>HGNC:18654</td>
      <td>Uncertain significance</td>
      <td>0</td>
      <td>Mar 18, 2020</td>
      <td>-1</td>
      <td>...</td>
      <td>criteria provided, single submitter</td>
      <td>1</td>
      <td>-</td>
      <td>N</td>
      <td>-</td>
      <td>2</td>
      <td>1177441</td>
      <td>70075478</td>
      <td>G</td>
      <td>A</td>
    </tr>
  </tbody>
</table>
<p>1535474 rows × 34 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">processBioDBs.utilities</span> <span class="kn">import</span> <span class="n">getSequence</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">easydict</span> <span class="kn">import</span> <span class="n">EasyDict</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">tqdm.notebook</span> <span class="kn">import</span> <span class="n">tqdm</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sequences</span> <span class="o">=</span> <span class="n">EasyDict</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">symbol</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">missense</span><span class="o">.</span><span class="n">GeneSymbol</span><span class="o">.</span><span class="n">unique</span><span class="p">()):</span>
    <span class="k">if</span> <span class="n">symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">sequences</span><span class="p">[</span><span class="n">symbol</span><span class="p">]</span> <span class="o">=</span> <span class="n">getSequence</span><span class="p">(</span><span class="n">symbol</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">continue</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

